lib/rucio/core/replica.py

```diff
@@ -1637,6 +1638,99 @@ def list_unlocked_replicas(rse, limit, bytes=None, rse_id=None, worker_number=No
+                             models.RSEFileAssociation.name,
+                             models.RSEFileAssociation.path,
+                             models.RSEFileAssociation.bytes,
+                             models.RSEFileAssociation.tombstone,
+                             models.RSEFileAssociation.state,
+                             models.RSEFileAssociation.rse_id).\
+        with_hint(models.RSEFileAssociation, "INDEX_RS_ASC(replicas REPLICAS_TOMBSTONE_IDX)  NO_INDEX_FFS(replicas REPLICAS_TOMBSTONE_IDX)", 'oracle').\
+        with_for_update(skip_locked=True).\
+        filter(models.RSEFileAssociation.tombstone < datetime.utcnow()).\
+        filter(models.RSEFileAssociation.lock_cnt == 0).\
+        filter(case([(models.RSEFileAssociation.tombstone != none_value, models.RSEFileAssociation.rse_id), ]) == rse_id).\
+        filter(or_(models.RSEFileAssociation.state.in_((ReplicaState.AVAILABLE, ReplicaState.UNAVAILABLE, ReplicaState.BAD)),
+                   and_(models.RSEFileAssociation.state == ReplicaState.BEING_DELETED, models.RSEFileAssociation.updated_at < datetime.utcnow() - timedelta(seconds=delay_seconds)))).\
+        order_by(models.RSEFileAssociation.tombstone).subquery()
+    query = session.query(subquery.c.scope,
+                          subquery.c.name,
+                          subquery.c.path,
+                          subquery.c.bytes,
+                          subquery.c.tombstone,
+                          subquery.c.state,
+                          func.count(models.RSEFileAssociation.scope)).\
+        join(models.RSEFileAssociation, and_(models.RSEFileAssociation.scope == subquery.c.scope,
+                                             models.RSEFileAssociation.name == subquery.c.name,
+                                             models.RSEFileAssociation.rse_id != subquery.c.rse_id,
+                                             models.RSEFileAssociation.state == ReplicaState.AVAILABLE)).\
+        group_by(subquery.c.scope, subquery.c.name, subquery.c.path, subquery.c.bytes, subquery.c.tombstone, subquery.c.state).\
+        order_by(subquery.c.tombstone)
+
+    needed_space = bytes
+    total_bytes, total_files = 0, 0
+    rows = []
+    for (scope, name, path, bytes, tombstone, state, cnt) in query.yield_per(1000):
+        # Check if more than one replica is available
+        if cnt > 1:
+            if state != ReplicaState.UNAVAILABLE:
+                total_bytes += bytes
+                if tombstone != OBSOLETE and needed_space is not None and total_bytes > needed_space:
+                    break
+
+                total_files += 1
+                if total_files > limit:
+                    break
+
+            rows.append({'scope': scope, 'name': name, 'path': path,
+                         'bytes': bytes, 'tombstone': tombstone,
+                         'state': state})
+            session.query(models.RSEFileAssociation).filter(models.RSEFileAssociation == scope,
```

Comment:
I think we should consider to do this one, and the one in the `else` condition in a bulk update too